### ISSUE TITLE
bug(linter) react/rules-of-hooks false positives

### ISSUE THREAD

**Issue by camc314** [oxc-project/oxc#7572]
# 1

```
  × eslint-plugin-react-hooks(rules-of-hooks): React Hook "useSWRFn" is called in function "Anonymous" that is neither a React function component nor a custom React Hook function. React component names must start with an uppercase letter. React Hook names must start with the word "use".
    ╭─[packages/frontend/core/src/hooks/use-query.ts:48:3]
 47 │       (immutable: boolean): useQueryFn =>
 48 │ ╭─▶   (options, config) => {
 49 │ │       const configWithSuspense: SWRConfiguration = useMemo(
 50 │ │         () => ({
 51 │ │           suspense: true,
 52 │ │           ...config,
 53 │ │         }),
 54 │ │         [config]
 55 │ │       );
 56 │ │   
 57 │ │       const useSWRFn = immutable ? useSWRImutable : useSWR;
 58 │ │       return useSWRFn(
 59 │ │         options ? () => ['cloud', options.query.id, options.variables] : null,
 60 │ │         options ? () => fetcher(options) : null,
 61 │ │         configWithSuspense
 62 │ │       );
 63 │ ╰─▶   };
 64 │     
    ╰────
```

https://github.com/toeverything/AFFiNE/blob/0ec1995addbb09fb5d4af765d84cc914b2905150/packages/frontend/core/src/hooks/use-query.ts#L46-L63

# ~~2~~ ✅ 

```
  × eslint-plugin-react-hooks(rules-of-hooks): React Hook "use" cannot be called at the top level. React Hooks must be called in a React function component or a custom React Hook function.
    ╭─[packages/common/infra/src/orm/core/validators/index.ts:45:24]
 44 │ export const validators = {
 45 │   validateTableSchema: use(tableSchemaValidators),
    ·                        ──────────────────────────
 46 │   validateCreateEntityData: use(createEntityDataValidators),
    ╰────
```

https://github.com/toeverything/AFFiNE/blob/a697ebe3401d5ad3afb7929a25f5e48a70488154/packages/common/infra/src/orm/core/validators/index.ts#L1-L50

This one is being correctly reported. since v5 when `use` hook was stabilised.

# ~~3~~ ✅ 
```
  × eslint-plugin-react-hooks(rules-of-hooks): React Hook "use" cannot be called at the top level. React Hooks must be called in a React function component or a custom React Hook function.
    ╭─[packages/backend/server/src/config/affine.self.ts:83:3]
 82 │ 
 83 │   AFFiNE.use('gcloud');
    ·   ────────────────────
 84 │ }
    ╰────
```

https://github.com/toeverything/AFFiNE/blob/638fc62601b58eb2a711f710e3cd80636af15798/packages/backend/server/src/config/affine.self.ts#L23

This one is being correctly reported. since v5 when `use` hook was stabilised.

# 4
```
  × eslint-plugin-react-hooks(rules-of-hooks): React Hook "useI18n" cannot be called inside a callback. React Hooks must be called in a React function component or a custom React Hook function.
    ╭─[packages/frontend/core/src/components/affine/affine-error-boundary/error-fallbacks/no-page-root-fallback.tsx:11:15]
 10 │     const { resetError } = props;
 11 │     const t = useI18n();
    ·               ─────────
 12 │ 
    ╰────
```

https://github.com/toeverything/AFFiNE/blob/7c0a686cd96bc1486fff6e154b7eb5bc7d8f49d0/packages/frontend/core/src/components/affine/affine-error-boundary/error-fallbacks/no-page-root-fallback.tsx#L7-L21
